### DISCUSSION TITLE
Interactions not trimming-friendly

### DISCUSSION THREAD

**Discussion by mysteryx93** [AvaloniaUI/Avalonia.Xaml.Behaviors#123] [General]
Using interactions breaks with Assembly Trimming. It's using reflection to access MediaFinished, and being defined on a 3rd party control, I won't be able to set attributes on that class.

```xaml
<Interaction.Behaviors>
    <EventTriggerBehavior EventName="MediaFinished">
        <InvokeCommandAction Command="{Binding PlayNext}" />
    </EventTriggerBehavior>
</Interaction.Behaviors>
```

I always thought that MethodBinding was a much better way to do it;  but couldn't get that to work in Avalonia.
https://github.com/mysteryx93/MethodBinding.Avalonia

What's the solution here?

**Comment by wieslawsoltes** [comment]
> Using interactions breaks with Assembly Trimming. It's using reflection to access MediaFinished, and being defined on a 3rd party control, I won't be able to set attributes on that class.
> 
> ```
> <Interaction.Behaviors>
>     <EventTriggerBehavior EventName="MediaFinished">
>         <InvokeCommandAction Command="{Binding PlayNext}" />
>     </EventTriggerBehavior>
> </Interaction.Behaviors>
> ```
> 
> I always thought that MethodBinding was a much better way to do it; but couldn't get that to work in Avalonia. https://github.com/mysteryx93/MethodBinding.Avalonia
> 
> What's the solution here?

Convert event as routed and use https://github.com/AvaloniaUI/Avalonia.Xaml.Behaviors/blob/62808959e9cff2c4d9022d71456845cb10aa2ad4/src/Avalonia.Xaml.Interactions.Custom/RoutedEventTriggerBehavior.cs#L11 
or create custom Behavior trigger.

**Reply by mysteryx93** [reply]
What's the difference with RoutedEventTriggerBehavior? That one doesn't rely on Reflection?

**Reply by mysteryx93** [reply]
OK I see, RoutedEvent property is set via a (compiled) binding. What's the recommended syntax to set that property to a RoutedEvent on the control though? Need to use some context-navigation syntax.

**Reply by mysteryx93** [reply]
It asks for a RoutedEvent; the RoutedEvent is static.

```c#
public static readonly RoutedEvent<RoutedEventArgs> StopCommandExecutedEvent =
    RoutedEvent.Register<MediaPlayerBase, RoutedEventArgs>(nameof(StopCommandExecuted), RoutingStrategies.Direct);
public event EventHandler<RoutedEventArgs>? StopCommandExecuted
{
    add => AddHandler(StopCommandExecutedEvent, value);
    remove => RemoveHandler(StopCommandExecutedEvent, value);
}
```

If I try to use like this

    <RoutedEventTriggerBehavior RoutedEvent="{x:Static ui:MediaPlayer.StopCommandExecutedEvent}">

I get

> MainView.axaml(131, 49): [AVLN:0004] Unable to resolve StopCommandExecutedEvent as static field, property, constant or enum value Line 131, position 49.

**Reply by mysteryx93** [reply]
Hah I was so close! This works. Need to call it on the base class.

    <RoutedEventTriggerBehavior RoutedEvent="{x:Static ui:MediaPlayerBase.StopCommandExecutedEvent}">
